game/view/actor.lua

```diff
@@ -88,186 +69,90 @@ function ActorView:draw()
-    if math.abs(unfocus_y-y) < 1 then y = unfocus_y end
-  else
-    y = math.round(y + (normal_y - y) * spd * dt)
-    if math.abs(normal_y-y) < 1 then y = normal_y end
-  end
-  self.importanthud_y = y
-
-  g.push()
-  g.setColor(COLORS.DARK)
-  g.print(pptext, 40, _height-y-fh*2)
-  g.print(xptext, 40, _height-y-fh)
-  if actor:getPrizePackCount() > 0 then
-    g.print(pcktext, 40, _height-y)
-  end
-  g.translate(-2, -2)
-  if actor:getPP() >= ACTIONDEFS.NEW_HAND_COST then
-    g.setColor(COLORS.SUCCESS)
-  else
-    g.setColor(COLORS.WARNING)
-  end
-  g.print(pptext, 40, _height-y-fh*2)
-  g.setColor(COLORS.NEUTRAL)
-  g.print(xptext, 40, _height-y-fh)
-  if actor:getPrizePackCount() > 0 then
-    g.setColor(COLORS.NOTIFICATION)
-    g.print(pcktext, 40, _height-y)
-  end
-  g.pop()
-
-  -- draw hand countdown
-  local handcountdown = actor:getHandCountdown()
-  local current = self.hand_count_down or 0
-  current = current + (handcountdown - current) * 0.2
-  if math.abs(current - handcountdown) < 1 then
-    current = handcountdown
-  end
-  self.hand_count_down = current
-  local handbar_percent = current / ACTIONDEFS.HAND_DURATION
-  local handbar_width = 492
-  local handbar_height = 12
-  g.push()
-  g.translate(40, _height - y + fh + handbar_height*2)
-  g.setLineWidth(1)
-  g.setColor(COLORS.BLACK)
-  g.rectangle('line', 0, 0, handbar_width/2, handbar_height)
-  g.setColor(COLORS.DARK)
-  g.rectangle('fill', 0, 0, handbar_width/2, handbar_height)
-  g.setColor(COLORS.WARNING)
-  g.rectangle('fill', 0, 0, handbar_width/2 * handbar_percent, handbar_height)
-  g.translate(0, -18)
-  local font = FONT.get("Text", 18)
-  font:set()
-  g.setColor(COLORS.BLACK)
-  g.print("Hand Duration", 0, 0)
-  g.translate(-1, -1)
+function ActorView:drawPanel(g)
   g.setColor(COLORS.NEUTRAL)
-  g.print("Hand Duration", 0, 0)
-  g.pop()
+  g.translate(3/4*g.getWidth(), 0)
+  ACTOR_PANEL.draw(g, -_PANEL_MG*2, -_PANEL_MG)
 end
 
 function ActorView:drawHP(g, actor)
-  g.push()
-  local cr, cg, cb = unpack(COLORS.NEUTRAL)
   local body = actor:getBody()
-  local hp, max_hp = body:getHP(), body:getMaxHP()
-  local str = _actor_text:format(hp, max_hp)
-  local w = _font:getWidth(str) + _font:getHeight()
-  g.translate(_width/2 - w/2, _height/2 + 20)
-  g.setColor(0, 0, 0, self.alpha)
-  g.printf(str, 0, 0, w, "center")
-  g.translate(-2, -2)
-  g.setColor(cr, cg, cb, self.alpha)
-  g.printf(str, 0, 0, w, "center")
-  g.pop()
+  local hp, pp = body:getHP(), actor:getPP()
+  local max_hp, max_pp = body:getMaxHP(), DEFS.MAX_PP
+  -- character name
+  FONT.set("TextBold", 24)
+  g.translate(_PANEL_MG, _PANEL_MG)
+  g.setColor(COLORS.NEUTRAL)
+  g.printf(actor:getTitle(), 0, -8, _PANEL_INNERWIDTH, "left")
+  -- character hp & pp
+  g.translate(0, 48)
+  ACTOR_HEADER.drawBar(g, "HP", hp, max_hp, COLORS.SUCCESS, COLORS.NOTIFICATION)
+  g.translate(0, 32)
+  ACTOR_HEADER.drawBar(g, "PP", pp, max_pp, COLORS.PP, COLORS.PP)
```

Comment:
If I understood right, this method purposefully leaves the current graphics transformation translated, which in turn influences things drawn thereafter. This seems to happen in a few other places as well.

Is it possible to write it in a way that uses `push()` and `pop()` so it's easier to modify the code in the future without messing with the whole layout?

(if it's too much, we can just create an issue for this instead)

Reply:
this way is better because each rendering section has its own relative positions, and i don't have to recalculate its base position at the beginning